test/acceptance_with_go_client/search_test.go

```diff
@@ -247,6 +248,21 @@ func TestHybridWithPureVectorSearch(t *testing.T) {
 	require.Len(t, result, 4)
 }
 
+func TestHybridWithNearTextSubsearch(t *testing.T) {
+	ctx := context.Background()
+	c := client.New(client.Config{Scheme: "http", Host: "localhost:8080"})
+	c.Schema().AllDeleter().Do(ctx)
+	className := "ParagraphWithManyWords"
+
+	AddClassAndObjects(t, className, string(schema.DataTypeTextArray), c, "text2vec-contextionary")
+	defer c.Schema().ClassDeleter().WithClassName(className).Do(ctx)
+
+	results, err := c.GraphQL().Raw().WithQuery(fmt.Sprintf("{Get{%s(hybrid: { nearText: {concepts: [\"rain\", \"nice\"]},  properties: [\"contents\"], alpha:1}, autocut: -1){num}}}", className)).Do(ctx)
```

Comment:
Didn't have time to look at it before, sorry that it took so long.

I am not sure if the new UX that comes with these changes is a good idea, I think it might be confusing. We now have four possibilities to add queries:
- hybrid: {query:""}
- hybrid: {query:"", vector: [...]}
- hybrid{near_text{...}}
- hybrid{nearVector{...}}

They all work differently and it is not clear to me which ones can be combined and which ones cannot and they have overlapping functionality.

Also if I add "vector" and "near_vector" I would expect a pure vector search. But what happens with near_text?

Reply:
You are right, I need to limit the combinations and provide feedback.  Could we put that in another patch, and merge this one before another patch breaks it?

(At the moment one vector search will take preference depending on the order of if statements in code)

Eventually I want to add a bm25 option for hybrid{nearVector{...}, bm25: {...}}.  Then the choices will be a simple query hybrid: {query:""} or your choice of subqueries.

I realise that just by looking at the patch, it seems complicated(and it is), but the alternatives are worse.  A custom query interface just for hybrid would require an entirely new set of options, with pages of documentation and instructions on how to combine query options.  Users can't have complete customisation and a simple interface.

If you have any specific ideas for improving it, please let me know

Reply:
There are some comments in the code about priority:

// If the user has given any weight to the vector search, choose 1 of three possible vector searches
	//
	// 1. If the user hase provided nearText parameters, use them in a nearText search
	// 2. If the user has provided nearVector parameters, use them in a nearVector search
	// 3. (Default) Do a vector search with the default parameters (the old hybrid search)

Reply:
I think my preference would be to only have one way of putting things in and then select the proper subsearch for the user.

eg use:
- query
- vector
- moveTo/moveAway

and based on the selection we can fill out nearText/nearVector behind the scenes. Or am I missing something?

Reply:
Sure, we can do that.  Users have asked for full customisation, so I've prepared a short list of the properties we will have to add to hybrid.  We can then parse each of these properties separately, write some code to combine them correctly into the original searches, then add all the new graphQL and GRPC to connect them.

Each time any programmer updates any other search in weaviate they just have to remember to duplicate all their work again inside hybrid, but slightly differently.  

Here's the properties:

* query
* vector
* distance
* certainity
* id
* beacon
* object_distance
* object_certainty
* concepts
* neartext_distance
* neartext_certainty
* autocorrect
* moveto
* moveto{concepts}
* moveTo{force}
* moveAwayFrom
* moveAwayFrom{concepts}
* moveAwayFrom{objects}
* moveAwayFrom{force}
* alpha
* properties
* fusiontype

Reply:
> They all work differently and it is not clear to me which ones can be combined and which ones cannot and they have overlapping functionality.

> Also if I add "vector" and "near_vector" I would expect a pure vector search. But what happens with near_text?

This problem will still exist with your proposal.  What happens when a user sends a search query, some concepts for neartext, and a vector?  How will we decide which combination of queries to run?

This is even less clear to me

Reply:
> We would have to expand this list when we add nearimage, nearaudio, etc.

yes, you have a point here. But I think current hybrid might simply not be the right place to add it.

> Here's the properties:

current parameters of hybrid:
- query
- vector
- alpha
- properties
- fusiontype

these make sense to add in my opinion (maybe with a different name to make clear that they only affect the vector-part):
- distance/certainty
- moveTo
- moveFrom

These not, either because they are just duplicates or don't make sense in the current hybrid search:
- id <= like adding near_object
- beacon <= same as id, unless we also wanted to support cross-class search (Explore), which is currently in an unusable state
- object_distance <= same as distance, just for near objects
- object_certainty <= same as certainty, just for near objects
- concepts <= same as query, supports inputting a list and if this is something users want we could accept both lists and strings for query - although I haven't seen anyone asking
- neartext_distance  <= same as distance, just for near text
- neartext_certaintyy <= same as certainty, just for near text
- autocorrect <= I dont even know what that does.

Just to give a few examples that I think will be hard to understand and hard to explain to users:
`hybrid: { nearText: {concepts: ["apple", "banana"]}}`
Does this do
 - A pure vector search
 - vector + keyword search
 
I think the first, but am pretty sure that the average user would guess the second.

`hybrid: { nearText: {concepts: ["apple", "banana"]}, properties: ["prop"]}`
Does this change anything to the above?

```
hybrid : {
      nearVector: {vector: [0.1, 0.2], targetVectors:"test"}, targetVectors:"test2"
    }
```
Which target vector is picked up? (I've read your comment about priority, so I'd assume the nearVector-targetVector is used)

```
hybrid : {
      nearVector: {vector: [0.1, 0.2], targetVectors:"test",}, nearText:{concepts:"abc"}
    }
```
Am I doing a hybrid search between `nearVector` and `nearText`? (I've read your comment, so I think this would just do a nearText-search)

We talked a while ago about true multi-search and I still really I like the idea. But this PR feels like this trying to push that concept into hybrid where it clashes with the current usage and creates foot-guns where the search is doing something unexpexted.

It might be the easiest from the engineering side but it will create a lot of effort for the docs team. I am happy to do/organize a session with them, happy to be wrong.

Reply:
I agree with all @dirkkul comments on the `hybrid` argument `UX`.
I also think that it would be great to have a `hybrid UX` session organized to talk about it.

Reply:
I think that current hybrid is actually very limited and underspecified, which is why I am arguing to make it much more specific.  I agree with your points about the logic and I should throw more errors to make things more clear.

My general response is that most of your points about confusion in my design also apply to your suggestions.  What happens if a user inputs neartext and nearvector in my method?  Exactly the same as your method.  We figure it out or throw an error when we can't.  We will use the exactly same logic (and share the same code, I suspect)

I disagree that my proposal is more work for the docs team, but please go ahead and organise a session.  Testing should probably be considered as well, since they will need to get the full specification in order to test that all the combinations work.

I can't help focussing on some points you made, so please excuse me while I respond directly to points.

> * A pure vector search
> * vector + keyword search
> I think the first, but am pretty sure that the average user would guess the second.

The answer is error, because there is no keyword search and alpha is not 1.

> hybrid: { nearText: {concepts: ["apple", "banana"]}, properties: ["prop"]}
> Does this change anything to the above?

Would it be better if I moved bm25 into it's on sub-query, so it is on the same level as the other sub-searches?  I was going to do it later.

> these make sense to add in my opinion (maybe with a different name to make clear that they only affect the vector-part):
> distance/certainty
> moveTo/moveFrom

These don't affect the whole vector part, they affect _different_ vector parts.  There is hidden information here. moveTo/moveFrom only applies to nearText.  Putting them in the top level implies they work for all vectors (to me, anyway), and they don't. This is what I feel is confusing about the current situation.  A user must carefully read the docs to know that moveTo can only be used on nearText.  And someone must write these docs.  And then test the feature.

> Which target vector is picked up? (I've read your comment about priority, so I'd assume the nearVector-targetVector is used)

Yes, subsearch wins.

> Am I doing a hybrid search between nearVector and nearText? (I've read your comment, so I think this would just do a nearText-search)

At the moment yes, but thinking about it, I think we should throw an error here too.

> autocorrect <= I dont even know what that does.

Same, that's why I want to just pass the entire configuration to the code that currently runs it.

Reply:
> I think that current hybrid is actually very limited and underspecified, which is why I am arguing to make it much more specific. I agree with your points about the logic and I should throw more errors to make things more clear.

I think the root of problem for me is that we can never get rid of the old way of doing things and then have two different usage-concepts next to each other. We need to support and explain both and this results in a lot of "if you are doing A you need to use Option X and Y but you cannot mix X and Z. However if you want to do B, you need to combine W and Y, but using X does not work".

What you are saying about the limitations/difficulties of hybrid makes total sense BUT I think need to go into a new endpoint (MultiSearch ?) that is designed from the group up to not have the confusing parts.

> These don't affect the whole vector part, they affect different vector parts. There is hidden information here. moveTo/moveFrom only applies to nearText. Putting them in the top level implies they work for all vectors (to me, anyway), and they don't. This is what I feel is confusing about the current situation. A user must carefully read the docs to know that moveTo can only be used on nearText. And someone must write these docs. And then test the feature.

Now you are saying it, yeah this is a problem. I feel like the simplest solution for _this_ problem would be to add support for move* to nearVector. I think someone, asked for it somewhere recently anyways.

Reply:
we found a good way to hide the complexity from users in the clients, this is resolved